client/src/lib.rs

```diff
@@ -476,28 +662,127 @@ impl<V: vdaf::Client<16>> Client<V> {
         let upload_endpoint = self
             .parameters
             .reports_resource_uri(&self.parameters.task_id)?;
-        let upload_response = retry_http_request(
+
+        #[cfg(feature = "ohttp")]
+        let upload_status = self.upload_with_ohttp(&upload_endpoint, &report).await?;
+        #[cfg(not(feature = "ohttp"))]
+        let upload_status = self.put_report(&upload_endpoint, &report).await?;
+
+        if !upload_status.is_success() {
+            return Err(Error::Http(Box::new(HttpErrorResponse::from(
+                upload_status,
+            ))));
+        }
+
+        Ok(())
+    }
+
+    async fn put_report(
+        &self,
+        upload_endpoint: &Url,
+        request_body: &[u8],
+    ) -> Result<StatusCode, Error> {
+        Ok(retry_http_request(
             self.parameters.http_request_retry_parameters.clone(),
             || async {
                 self.http_client
                     .put(upload_endpoint.clone())
                     .header(CONTENT_TYPE, Report::MEDIA_TYPE)
-                    .body(report.clone())
+                    .body(request_body.to_vec())
                     .send()
                     .await
             },
         )
-        .await
-        .map_err(|err| match err {
-            Ok(http_error_response) => Error::Http(Box::new(http_error_response)),
-            Err(error) => error.into(),
-        })?;
+        .await?
+        .status())
+    }
+
+    /// Send a DAP upload request via OHTTP. This is only intended for DAP uploads and so does not
+    /// handle response bodies.
+    #[cfg(feature = "ohttp")]
+    #[tracing::instrument(skip(self), err)]
+    async fn upload_with_ohttp(
+        &self,
+        upload_endpoint: &Url,
+        request_body: &[u8],
+    ) -> Result<StatusCode, Error> {
+        let (ohttp_config, key_configs) =
+            if let Some((ohttp_config, key_configs)) = &self.ohttp_config {
+                (ohttp_config, key_configs)
+            } else {
+                return self.put_report(upload_endpoint, request_body).await;
+            };
```

Comment:
IIUC if `ohttp_config` is None, then we fallback to normal uploads? This strikes me as a footgun for an accidental privacy leak if the client is used improperly.

I understand this as allowing non-OHTTP uploads with the `ohttp` feature enabled`. The point above about having a separate suite of methods would resolve this.

Reply:
I don't see it as a fallback. If the client didn't ask for OHTTP, we don't do OHTTP.

You're saying it should be impossible to do non-OHTTP uploads if feature `ohttp` is enabled? I don't agree. Case in point: we use `janus_client` in our own integration tests and we want to be able to exercise both OHTTP and non-OHTTP uploads without recompiling.

Reply:
I see, I think my problem then is moreso code organization--it raises my alarm bells that we could do non-OHTTP upload in a function called `upload_with_ohttp`. Maybe hoisting this logic up to `upload_with_time()` would make it clearer.

Can take it or leave it, I don't feel strongly about it now that I know the intent.

> You're saying it should be impossible to do non-OHTTP uploads if feature ohttp is enabled?

Not really, I'm moreso looking for explicitness about whether an upload is being performed with OHTTP or not. Which, eh, I think `.with_ohttp_config()` in the client builder is sufficient, so no action necessary.

Reply:
This is the least stupid way I found to do this. It's very awkward to work around the `ohttp` feature. I added a comment to this method explaining that it might not do OHTTP, but since it's not a public item, I'm less concerned about misuse.